---
title: "IA-SVA based feature selection improves the performance of clustering algorithms [1]"
author: "Donghyung Lee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The IA-SVA based feature selection can significantly improve the performance and utility of clustering algorithms (e.g., tSNE, hierarchical clustering). To illustrate how the IA-SVA method can be used to improve the performance of clustering algorithms, we used real-world single cell RNA sequencing (scRNA-Seq) data obtained from human pancreatic islet samples ([Lawlor et. al., 2016](http://genome.cshlp.org/content/early/2017/01/16/gr.212720.116)). This dataset is included in a R data package ("iasvaExamples") containing data examples for IA-SVA (https://github.com/dleelab/iasvaExamples). To install the 'iasvaExamples' package, follow the instruction provided in the GitHub page.

## Load packages
```{r load_packages, echo=TRUE}
rm(list=ls())
library(irlba)
library(iasva)
library(iasvaExamples)
library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson's contingency coefficient
library(RColorBrewer)

color.vec <- brewer.pal(8, "Set1")
```

## Load the islet single cell RNA-Seq data (n=638 cells, and 26K genes)
```{r load_data, echo=TRUE}
data("Lawlor_Islet_scRNAseq_Read_Counts")
data("Lawlor_Islet_scRNAseq_Annotations")
ls()
counts <- Lawlor_Islet_scRNAseq_Read_Counts
anns <- Lawlor_Islet_scRNAseq_Annotations
dim(anns)
dim(counts)

summary(anns)
ContCoef(table(anns$Gender, anns$Cell_Type))
ContCoef(table(anns$Phenotype, anns$Cell_Type))
ContCoef(table(anns$Race, anns$Cell_Type))
ContCoef(table(anns$Patient_ID, anns$Cell_Type))
ContCoef(table(anns$Batch, anns$Cell_Type))
```

The annotations describing the islet samples and experimental settings are stored in "anns" and the read counts information is stored in "counts". 

## Extract three cell types (GCG (alpha), INS (beta), KRT19 (ductal) expressing cells) from healthy (i.e., non-diabetic) subjects and filter out low-expressed genes 
```{r alpha_cells, echo=TRUE, results='asis'}
counts <- counts[, (anns$Phenotype!="Non-Diabetic")&
                    ((anns$Cell_Type=="GCG")|
                      (anns$Cell_Type=="INS")|
                        (anns$Cell_Type=="KRT19"))]
anns <- subset(anns, (Phenotype!="Non-Diabetic")& 
                      ((Cell_Type=="GCG")|
                        (Cell_Type=="INS")|
                          (Cell_Type=="KRT19")))
dim(counts)
dim(anns)

anns <- droplevels(anns)

prop.zeros <- sum(counts==0)/length(counts)
prop.zeros

filter = apply(counts, 1, function(x) length(x[x>5])>=3)

counts = counts[filter,]
dim(counts)

prop.zeros <- sum(counts==0)/length(counts)
prop.zeros

Patient_ID <- anns$Patient_ID
Cell_Type <- anns$Cell_Type
Batch <- anns$Batch
```

## Calculate geometric library size, i.e., library size of log-transfromed read counts 
It is well known that the geometric library size (i.e., library size of log-transfromed read counts) or proportion of expressed genes in each cell explains a very large portion of variability in scRNA-Seq data ([Hicks et. al. 2015 BioRxiv](http://biorxiv.org/content/early/2015/08/25/025528), [McDavid et. al. 2016 Nature Biotechnology](http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html)). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the geometric library size (e.g., r > 0.9). Here, we calculate the geometric library size for islet cells, which will be used as an known factor in the IA-SVA analyses. 
```{r geno_lib_size, echo=TRUE, fig.width=7, fig.height=4}
Geo_Lib_Size <- colSums(log(counts+1))
barplot(Geo_Lib_Size, xlab="Cell", las=2, ylab = "Geometric Library Size")
lcounts <- log(counts + 1)

# PC1 and Geometric library size correlation
pc1 = irlba(lcounts - rowMeans(lcounts), 1)$v[,1] ## partial SVD
cor(Geo_Lib_Size, pc1)
```

## Run tSNE to cluster islet cells.
For comparison purposes, we applied tSNE on read counts of all genes. We used the Rtsne R package with default settings for this analyses.
Genes are colored with respect to the expression of marker genes. 
```{r run_tsne, echo=TRUE, fig.width=7, fig.height=7}
set.seed(323542534)
tsne.res <- Rtsne(t(lcounts), dims = 2)

plot(tsne.res$Y, main="tSNE", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend("bottomright", levels(Cell_Type), border="white",fill=color.vec, bty="n")
```

## Run IA-SVA
Here, we first run IA-SVA using Patient_ID, Batch and Geo_Lib_Size as known factors and identify 5 hidden factors.
Since cell type is not used as a known factor in this analyses, IA-SVA will detect the heterogeneity associated with the cell types.
```{r run_iasva, echo=TRUE, fig.width= 7, fig.height=6}
mod <- model.matrix(~Patient_ID+Batch+Geo_Lib_Size)
iasva.res<- iasva(t(counts), mod[,-1],verbose=FALSE, permute=FALSE, num.sv=5)
iasva.sv <- iasva.res$sv

#with color-coding based on true cell-type
pairs(iasva.sv, main="IA-SVA", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], oma=c(4,4,6,14))
legend("right", levels(Cell_Type), border="white", fill=color.vec, bty="n")

cor(Geo_Lib_Size, iasva.sv)

corrplot(cor(iasva.sv))
```

As shown in the above figure, the fist two SVs (SV1 and SV2) separate cell types.

## Find marker genes for SV1 and SV2.
Here, using the find.markers() function we find marker genes significantly associated with SV1 and SV2 (multiple testing adjusted p-value < 0.05, default significance cutoff, a high R-squared value: R-squared > 0.4).
```{r find_markers, echo=TRUE, fig.width=7, fig.height=10}
marker.counts <- find.markers(t(counts), as.matrix(iasva.sv[,c(1,2)]), rsq.cutoff = 0.4)
nrow(marker.counts)

anno.col <- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) <- colnames(marker.counts)
head(anno.col)

cell.type.col <- color.vec[1:3]
names(cell.type.col) <- c("GCG","INS","KRT19")
anno.colors <- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts+1), show_colnames =FALSE, fontsize_row = 7, clustering_method = "ward.D2", cutree_cols = 3, annotation_col = anno.col, annotation_colors = anno.colors)
```

This list include 86 genes including INS, GCG, and KRT19. In the case of islet cells, marker genes are well established and IA-SVA did an excellent job of redefining these markers along with some other highly informative genes. Therefore, IA-SVA can be effectively used to uncover heterogeneity associated with cell types and can reveal genes that are expressed in a cell-specific manner.

## Run tSNE post IA-SVA, i.e., run tSNE on marker genes for SV1 and SV2 obtained from IA-SVA.
Here, we apply tSNE on the marker genes for SV1 and SV2 
```{r run_tsne_post_iasva, echo=TRUE, fig.width=7, fig.height=7}
set.seed(3445462)
tsne.res <- Rtsne(unique(t(log(marker.counts+1))), dims = 2)

plot(tsne.res$Y, main="tSNE post IA-SVA", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend("bottomright", levels(Cell_Type), border="white", fill=color.vec, bty="n")
```

tSNE conducted on genes selected via IA-SVA very clearly seperates cells into their corresponding clusters. Moreover, this analyses also revealed
one cell (green cell clustered together with blue cells) that is potentially mislabeled in the original analyses. 


## Session Info
```{r session_info, echo=TRUE}
sessionInfo()
```
