---
title: "IA-SVA based feature selection improves the performance of clustering algorithms [2]"
author: "Donghyung Lee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Here, we used single cell RNA sequencing (scRNA-Seq) data with strong confounding variables, which is also obtained from human pancreatic islet samples ([Xin et. al., 2016](http://www.cell.com/cell-metabolism/abstract/S1550-4131(16)30434-X)). This dataset is included in an R data package ("iasvaExamples") containing data examples for IA-SVA (https://github.com/dleelab/iasvaExamples). To install the 'iasvaExamples' package, follow the instructions provided in the GitHub page.

## Load packages
```{r load_packages, echo=TRUE}
rm(list=ls())
library(iasva)
library(iasvaExamples)
library(sva)
library(irlba)
library(Rtsne)
library(pheatmap)
library(corrplot)
library(DescTools) #pcc i.e., Pearson's contingency coefficient
library(RColorBrewer)

color.vec <- brewer.pal(8, "Set1")

#color.pal from https://www.r-bloggers.com/the-paul-tol-21-color-salute/
tol21rainbow= c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788")
```

## Load the islet scRNA-Seq data (638 genes, >26K genes)
```{r load_data, echo=TRUE, fig.width=6, fig.height=4}
data("Xin_Islet_scRNAseq_Read_Counts")
data("Xin_Islet_scRNAseq_Annotations")
ls()
counts <- Xin_Islet_scRNAseq_Read_Counts
anns <- Xin_Islet_scRNAseq_Annotations
dim(anns)
dim(counts)

Lib_Size <- colSums(counts)
plot(sort(Lib_Size))
hist(Lib_Size)
summary(Lib_Size)
```
The annotations describing the islet samples and experimental settings are stored in "anns" and the read counts information is stored in "counts". 

## Filter out low-expressed genes dataset reduced to 18K genes and 617 cells
```{r alpha_cells, echo=TRUE, results='asis'}
##counts <- counts[, (anns$Cell_Type!="none")] 
##anns <- subset(anns, (Cell_Type!="none"))
dim(counts)
dim(anns)

anns <- droplevels(anns)

prop.zeros <- sum(counts==0)/length(counts)
prop.zeros

filter = apply(counts, 1, function(x) length(x[x>5])>=3)

counts = counts[filter,]
dim(counts)

prop.zeros <- sum(counts==0)/length(counts)
prop.zeros

summary(anns)

Patient_ID <- anns$Donor_ID
Gender <- anns$Gender
Age <- anns$Age
Cell_Type <- anns$Cell_Type
Phenotype <- anns$Condition
Ethnicity <- anns$Ethnicity
Mito_Frac <- anns$Mitochondrial_Fraction

ContCoef(table(Cell_Type, Patient_ID))
ContCoef(table(Cell_Type, Gender))
ContCoef(table(Cell_Type, Age))
ContCoef(table(Cell_Type, Phenotype))
ContCoef(table(Cell_Type, Ethnicity))
```

Note that the orignial cell assignments are highly correlated with known factors (e.g., Patient_ID (pcc=0.48)).

## Calculate geometric library size, i.e., library size of log-transfromed read counts 
It is well known that the geometric library size (i.e., library size of log-transfromed read counts) or proportion of expressed genes in each cell explains a very large portion of variability of scRNA-Seq data ([Hicks et. al. 2015 BioRxiv](http://biorxiv.org/content/early/2015/08/25/025528), [McDavid et. al. 2016 Nature Biotechnology](http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html)). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the geometric library size (r > 0.9). Here, we calculate the geometric library size to be used as a known factor in the IA-SVA analyses.
```{r geno_lib_size, echo=TRUE, fig.width=6, fig.height=4}
Geo_Lib_Size <- colSums(log(counts+1))
barplot(Geo_Lib_Size, xlab="Cell", las =2)
lcounts <- log(counts + 1)
pca.res = irlba(lcounts - rowMeans(lcounts), 5)$v
cor(Geo_Lib_Size, pca.res[,1])
dim(lcounts)
```

## Run tSNE to cluster islet cells.
For comparison purposes,we applied tSNE on read counts of all genes. We used the Rtsne R package with default settings.
Genes are color coded wrt their expression of marker genes. 
```{r run_tsne, echo=TRUE, fig.width=7, fig.height=7}
set.seed(34544532)
tsne.res.all <- Rtsne(t(lcounts), dims = 2)

plot(tsne.res.all$Y, main="tSNE", xlab="tSNE Dim1", ylab="tSNE Dim2",pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend("topright", levels(Cell_Type), border="white", fill=color.vec, bty="n")

par(mfrow=c(2,2))
plot(tsne.res.all$Y, main="Gender", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=color.vec[Gender], bg=color.vec[Gender], oma=c(4,4,6,12))
legend("topright", levels(Gender), border="white", fill=color.vec, bty="n")
plot(tsne.res.all$Y, main="Patient ID", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=tol21rainbow[Patient_ID], bg=tol21rainbow[Patient_ID], oma=c(4,4,6,12))
legend("topright", levels(Patient_ID), border="white", fill=tol21rainbow, bty="n", cex=0.5)
plot(tsne.res.all$Y, main="Ethnicity", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=color.vec[Ethnicity], bg=color.vec[Ethnicity], oma=c(4,4,6,12))
legend("topright", levels(Ethnicity), border="white", fill=color.vec, bty="n")
plot(tsne.res.all$Y, main="Phenotype", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=color.vec[Phenotype], bg=color.vec[Phenotype], oma=c(4,4,6,12))
legend("topright", levels(Phenotype), border="white", fill=color.vec, bty="n")
par(mfrow=c(1,1))
```

Known factors deteriorate the performance of t-SNE.


## Run PCA to cluster islet cells.
Here, we applied PCA on read counts of all genes. Genes are color coded wrt their expression of marker genes. 
```{r run_pca, echo=TRUE, fig.width=7, fig.height=6}
pairs(pca.res[,1:4], main="PCA", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], cex=0.8, oma=c(4,4,6,12))
legend("right", levels(Cell_Type), border="white", fill=color.vec, bty="n")
```

## Run surrogate variable analysis (SVA).
Here, for comparison purposes we conduct SVA on our example data while adjusting for Patient_ID and Geo_Lib_Size and obtained 4 hidden factors. 
```{r run_sva, echo=TRUE, eval=TRUE, fig.width=7, fig.height=6}
mod1 <- model.matrix(~Patient_ID+Geo_Lib_Size)
mod0 <- cbind(mod1[,1])
sva.res = svaseq(counts,mod1,mod0, n.sv=4)$sv

pairs(sva.res[,1:4], main="SVA", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], cex=0.8, oma=c(4,4,6,12))
legend("right", levels(Cell_Type), border="white", fill=color.vec, bty="n")
```

## Run IA-SVA
Here, we run IA-SVA using Patient_ID and Geo_Lib_Size as known factors and identify 4 hidden factors. 
```{r run_iasva, echo=TRUE, fig.width=7, fig.height=6}
mod <- model.matrix(~Patient_ID+Geo_Lib_Size)
iasva.res<- iasva(t(counts), mod[,-1],verbose=FALSE, permute=FALSE, num.sv=4)
iasva.sv <- iasva.res$sv

## no color
pairs(iasva.sv[,1:4], pch=21, col="black", bg="black", cex=0.8)

## with color-coding
pairs(iasva.sv[,1:4], main="IA-SVA", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], cex=0.8, oma=c(4,4,6,12))
legend("right", levels(Cell_Type), border="white", fill=color.vec, bty="n")
```

As shown in the above figure (with no color), SV1, SV3 and SV4 separate cells. Since SV3 captures cell contamination, we consider S1 and SV4 in the subsequent analysis.

## Correlation between SVs
```{r corr_btw_SVs, echo=TRUE, fig.width=6, fig.height=4}
cor(iasva.sv)
corrplot(cor(iasva.sv))
```

## Find marker genes for SV1 and SV4.
Here, we find marker genes significantly associated with SV1 and SV4 (multiple testing adjusted p-value < 0.05, default significance cutoff) and having a high R-squared value (R-squared > 0.3) using the find.markers() function.
```{r find_markers, echo=TRUE, fig.width=7, fig.height=10}
marker.counts <- find.markers(t(counts), as.matrix(iasva.sv[,c(1,4)]),  rsq.cutoff = 0.3)
nrow(marker.counts)

anno.col <- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) <- colnames(marker.counts)
head(anno.col)
cell.type.col <- color.vec[1:4]
names(cell.type.col) <- c("alpha","beta","delta","PP")
anno.colors <- list(Cell_Type=cell.type.col)

pheatmap(log(marker.counts+1), show_colnames =FALSE, show_rownames = TRUE, clustering_method = "ward.D2",cutree_cols = 5,annotation_col = anno.col, annotation_colors=anno.colors)
```

## Run tSNE post IA-SVA, i.e., run tSNE on 57 marker genes obtained from IA-SVA.
Here, we apply tSNE on the marker genes for SV1 and SV4 obtained from IA-SVA
```{r run_tsne_post_iasva, echo=TRUE, fig.width=7, fig.height=7}
set.seed(75458456)
tsne.res <- Rtsne(unique(t(log(marker.counts+1))), dims = 2)

plot(tsne.res$Y, main="tSNE post IA-SVA", xlab="tSNE Dim1", ylab="tSNE Dim2", pch=21, col=color.vec[Cell_Type], bg=color.vec[Cell_Type], oma=c(4,4,6,12))
legend("bottomright", levels(Cell_Type), border="white", fill=color.vec, bty="n")
```

tSNE performed on marker genes selected via IA-SVA performs better than original tSNE analyses using all genes. 
This example again reiterates the importance of gene selection using IA-SVA for effective clustering of single-cell datasets.


## Session Info
```{r session_info, echo=TRUE}
sessionInfo()
```
