---
title: "Detecting hidden heterogeneity in single cell RNA-Seq data"
author: "Donghyung Lee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


The iasva package can be used to detect hidden heterogenity within bulk or single cell sequencing data. To illustrate how to use the iasva package for heterogenity detection, we use real-world single cell RNA sequencing (scRNA-Seq) data obtained from human pancreatic islet samples ([Lawlor et. al., 2016](http://genome.cshlp.org/content/early/2017/01/16/gr.212720.116)). This dataset is included in a R data package ("iasvaExamples") containing data examples for IA-SVA (https://github.com/dleelab/iasvaExamples). To install the package, follow the instruction provided in the GitHub page.

## Load packages
```{r load_packages, echo=TRUE}
rm(list=ls())
library(iasva)
library(iasvaExamples)
library(sva)
library(Rtsne)
library(pheatmap)
```

## Load the islet single cell RNA-Seq data
```{r load_data, echo=TRUE}
data("Lawlor_Islet_scRNAseq_Read_Counts")
data("Lawlor_Islet_scRNAseq_Annotations")
ls()
counts <- Lawlor_Islet_scRNAseq_Read_Counts
anns <- Lawlor_Islet_scRNAseq_Annotations
dim(anns)
dim(counts)
# set seed for the random number generator.
set.seed(344532456)
```

The annotations describing the islet samples and experimental settings are stored in "anns" and the read counts information is stored in "counts". 

## Extract alpha cells (GCG expressed cells) from non-diabetics 
To illustrate how IA-SVA can be used to detect hidden heterogeneity within a homogenous cell population, we use read counts of alpha cells from non-diabetic subjects.   
```{r alpha_cells, echo=TRUE, results='asis'}
counts <- counts[, (anns$Phenotype!="Non-Diabetic")&(anns$Cell_Type=="GCG")] 
anns <- subset(anns, (Phenotype!="Non-Diabetic")&(Cell_Type=="GCG"))
dim(counts)
dim(anns)

anns <- droplevels(anns)

prop.zeros <- sum(counts==0)/length(counts)
prop.zeros

filter = apply(counts, 1, function(x) length(x[x>5])>=3)

counts = counts[filter,]
dim(counts)

prop.zeros <- sum(counts==0)/length(counts)
prop.zeros
```

## Calculate geometric library size, i.e., library size of log-transfromed read counts 
It is well known that the geometric library size (i.e., library size of log-transfromed read counts) or proportion of expressed genes in each cell explains very large portion of variability of scRNA-Seq data ([Hicks et. al. 2015 BioRxiv](http://biorxiv.org/content/early/2015/08/25/025528), [McDavid et. al. 2016 Nature Biotechnology](http://www.nature.com/nbt/journal/v34/n6/full/nbt.3498.html)). Frequently, the first principal component of log-transformed scRNA-Seq read counts is highly correlated with the geometric library size. (r > 0.9). Here, we calculate it. Later this vector will be used as an known factor in IA-SVA algorithm.
```{r geno_lib_size, echo=TRUE, fig.width=6, fig.height=4}
Geo_Lib_Size <- colSums(log(counts+1))
barplot(Geo_Lib_Size, xlab="Cell")
lcounts <- log(counts + 1)
pca.res = svd(lcounts - rowMeans(lcounts))$v
cor(Geo_Lib_Size, pca.res[,1])
```

## Run IA-SVA
Here, we run IA-SVA using Patient_ID and Geo_Lib_Size as known factors and identify hidden factors. 
```{r run_iasva, echo=TRUE, fig.width= 6, fig.height=5}
Patient_ID <- anns$Patient_ID
mod <- model.matrix(~Patient_ID+Geo_Lib_Size)
iasva.res<- iasva(t(counts), mod[,-1],verbose=FALSE, permute=FALSE, num.sv=3)
iasva.sv <- iasva.res$sv
plot(iasva.sv[,1], iasva.sv[,2], xlab="SV1", ylab="SV2")

Cell_Type <- as.factor(iasva.sv[,2] > -0.2) 
levels(Cell_Type)=c("Cell1","Cell2")

pairs(iasva.sv, main="IA-SVA", pch=21, bg=c("red","green3")[Cell_Type], oma=c(4,4,6,12)) #4,4,6,12
legend(0.8, 0.55, levels(Cell_Type), fill=c("red", "green3"), bty="n")
```

As shown in the above figure, SV2 clearly separates alpha cells into two groups (6 outlying cells and the rest of cells). SV3 also captures an outlying cell but let's focus on SV2.


## Find marker genes for the detected heterogeneity (SV2).
Here, we find marker genes significantly associated with SV2 (multiple testing adjusted p-value < 0.05, default significance cutoff) and having a high R-squared value (R-squared > 0.3, default R-squared cutoff) using the find.markers() function.
```{r find_markers, echo=TRUE, fig.width=8, fig.height=12}
marker.counts <- find.markers(t(counts), as.matrix(iasva.sv[,2]))
nrow(marker.counts)

anno.col <- data.frame(Cell_Type=Cell_Type)
rownames(anno.col) <- colnames(marker.counts)
head(anno.col)
pheatmap(log(marker.counts+1), show_colnames =FALSE, clustering_method = "ward.D2",cutree_cols = 2,annotation_col = anno.col)
```

## Run tSNE to detect the hidden heterogeneity (SV2).
For comparison purpose, here we applied tSNE on read counts of all genes to identify the hidden heterogeneity. We used the Rtsne R package with default settings.

```{r run_tsne, echo=TRUE, fig.width=6, fig.height=5}
tsne.res <- Rtsne(t(lcounts), dims = 2)
pairs(tsne.res$Y, main="tSNE", pch=21, bg=c("red","green3")[Cell_Type], oma=c(4,4,6,12)) #4,4,6,12
legend(0.8, 0.55, levels(Cell_Type), fill=c("red", "green3"), bty="n")
```

As shown above, tSNE couldn't detect the outlying cells identified by IA-SVA

## Run tSNE post IA-SVA, i.e., run tSNE on maker genes for SV2 obtained from IA-SVA.
Here, we apply tSNE on the marker genes for SV2 obtained from IA-SVA
```{r run_tsne_post_iasva, echo=TRUE, fig.width=6, fig.height=5}
tsne.res <- Rtsne(unique(t(log(marker.counts+1))), dims = 2)
pairs(tsne.res$Y, main="tSNE post IA-SVA", pch=21, bg=c("red","green3")[Cell_Type], oma=c(4,4,6,12)) #4,4,6,12
legend(0.8, 0.55, levels(Cell_Type), fill=c("red", "green3"), bty="n")
```

tSNE post IA-SVA perform better than regular tSNE applied on read counts of entire genes. 


## Run principal component analysis (PCA) to detect the hidden heterogeneity (SV2).
Here, we use PCA to detect the hidden heterogeneity detected by IA-SVA. 
```{r run_pca, echo=TRUE, fig.width=6, fig.height=5}
pca_res = svd(lcounts - rowMeans(lcounts))$v
pairs(pca.res[,1:3], main="PCA", pch=21, bg=c("red","green3")[Cell_Type], oma=c(4,4,6,12)) #4,4,6,12
legend(0.8, 0.55, levels(Cell_Type), fill=c("red", "green3"), bty="n")
```

Separation of the six outlying cells by var3 (PC3) is not as clear as one observed from IA-SVA results.



## Run surrogate variable analysis (SVA) to detect the hidden heterogeneity (SV2).
Here, we use SVA to detect the hidden heterogeneity detected by IA-SVA. 
```{r run_sva, echo=TRUE, fig.width=6, fig.height=5}
mod1 <- model.matrix(~Patient_ID+Geo_Lib_Size)
mod0 <- cbind(mod1[,1])

sva.res = svaseq(counts,mod1,mod0, n.sv=3)$sv
##plot(sva.res[,1], sva.res[,2], xlab="SV1", ylab="SV2", col=Cell_Type)
pairs(sva.res[,1:3], main="SVA", pch=21, bg=c("red","green3")[Cell_Type], oma=c(4,4,6,12)) #4,4,6,12
legend(0.8, 0.55, levels(Cell_Type), fill=c("red", "green3"), bty="n")
```

Again, separation of the six outlying cells by var2 (SV2) is not as clear as one from IA-SVA.


## Session Info
```{r session_info, echo=TRUE}
sessionInfo()
```
